### PR TITLE
adding compatibility to Spark 3

### LINKED ISSUE #92: Add support for Databricks Runtime 7.0 (Scala 2.12, Spark 3.0)
Please add support for the new Databricks Runtime 7.0 which is based on Scala 2.12 and Spark 3.0. 

Adding a maven profile for it to **pom.xml** does not work because of too many changes in the new Spark and Scala versions:
```xml
<profile>
            <id>scala-2.12_spark-3.0.0</id>
            <activation>
                <activeByDefault>true</activeByDefault>
            </activation>
            <properties>
                <codahale.metrics.version>3.1.5</codahale.metrics.version>
                <commons-codec.version>1.10</commons-codec.version>
                <commons.httpclient.version>4.5.6</commons.httpclient.version>
                <fasterxml.jackson.databind.version>2.10.0</fasterxml.jackson.databind.version>
                <jetty.version>9.4.18.v20190429</jetty.version>
                <log4j.version>1.2.17</log4j.version>
                <slf4j.version>1.7.30</slf4j.version>
                <spark.version>3.0.0</spark.version>
                <scala.version>2.12.10</scala.version>
                <scala.compat.version>2.12</scala.compat.version>
            </properties>
        </profile>
```

### PR DESCRIPTION
**Pull request by dfanesiDB**
### Changes proposed in this PR
I have added the support for spark 3 using the same codebase and same tests for both versions of spark and scala. 

### Code changes:
StreamingListenerHandlers.scala:
It looks like a method was renamed in the new release of spark, so I am taking away the optional "override" keyword and adding the new method as well. 

### Changes to unit tests
The new version of the io.dropwizard.metrics:metrics-core has moved the CpuTimeClock class. The only way to support both versions with the same tests is to instantiate the object using reflections. 

Spark 3 has also added some new parameters to the constructors of events. I decided to use the same approach to not duplicate code: I am instantiating the objects of those classes using reflections.

Reflections add a lot of complexity to the unit tests, but it enables us to reuse the same unit tests and to not duplicate code. The alternative would be to duplicate code for the unit tests, but then we would have to maintain two unit test suites, which is not ideal as well. Unfortunately, none of the choices are good in my opinion. 

Fix #92

### CODE REVIEW

**Comment by Nemeczek** [comment]
I hope this will get merged soon :)

**Comment by hallihan** [comment]
@dfanesiDB - are you able to review/sign the CLA linked above?

**Comment by dfanesiDB** [comment]
Hi @hallihan, unfortunately, I am still waiting the approval of my legal team. I will ping them again today to see if they are ok with me signing this. I guess you can start the review in the meantime if you want to.

**Comment by hallihan** [comment]
@dfanesiDB any update on the CLA?  I'm not able to start a review until this is signed.

**Comment by dfanesiDB** [comment]
@hallihan sorry for the delay. I was finally able to get the approval I needed.

**Comment by hallihan** [comment]
> @hallihan sorry for the delay. I was finally able to get the approval I needed.

Great!  I should be able to look at this over the next couple of days.

File: README.md [hunk 1]
```diff
@@ -144,6 +144,8 @@ databricks runtime.
 | -- | -- |
 | `5.5` | `scala-2.11_spark-2.4.3` |
 | `6.4` - `6.6` | `scala-2.11_spark-2.4.5` |
+| `7.0` - `7.2` | `scala-2.12_spark-3.0.0` |
+| `7.3` | `scala-2.12_spark-3.0.1` |
```
Comment (hallihan) [inline README.md:148]:
```suggestion
| `7.3` - `7.4` | `scala-2.12_spark-3.0.1` |
```

File: src/pom.xml [hunk 2]
```diff
@@ -37,6 +37,32 @@
                 <scala.compat.version>2.11</scala.compat.version>
             </properties>
         </profile>
+        <profile>
+            <id>scala-2.12_spark-3.0.0</id>
+            <activation>
```
Comment (hallihan) [inline src/pom.xml:42]: Only one profile should be marked `activeByDefault`, can you remove it from `scala-2.11_spark-2.4.5` and `scala-2.12_spark-3.0.0`?

File: src/pom.xml [hunk 3]
```diff
@@ -37,6 +37,32 @@
                 <scala.compat.version>2.11</scala.compat.version>
             </properties>
         </profile>
+        <profile>
```
Comment (hallihan) [inline src/pom.xml:40]: Adding comment here since GitHub won't let me comment on files that are outside of the diff.  Can you add profiles for `scala-2.12_spark-3.0.0` and `scala-2.12_spark-3.0.1` in `sample/spark-sample-job/pom.xml`?  We tell the user in the readme to use these same profiles when building the sample and if the profile is missing they will get a warning like:
> [WARNING] The requested profile "scala-2.12_spark-3.0.1" could not be activated because it does not exist.
and the build will proceed with the `activeByDefault` profile.

See the diff below for the changes I made when testing:
```
diff --git a/sample/spark-sample-job/pom.xml b/sample/spark-sample-job/pom.xml
index adfea2e..ee71f7c 100644
--- a/sample/spark-sample-job/pom.xml
+++ b/sample/spark-sample-job/pom.xml
@@ -21,9 +21,6 @@
         </profile>
         <profile>
             <id>scala-2.11_spark-2.4.5</id>
-            <activation>
-                <activeByDefault>true</activeByDefault>
-            </activation>
             <properties>
                 <jetty.version>9.4.31.v20200723</jetty.version>
                 <slf4j.version>1.7.16</slf4j.version>
@@ -32,6 +29,29 @@
                 <scala.compat.version>2.11</scala.compat.version>
             </properties>
         </profile>
+        <profile>
+            <id>scala-2.12_spark-3.0.0</id>
+            <properties>
+                <jetty.version>9.4.31.v20200723</jetty.version>
+                <slf4j.version>1.7.16</slf4j.version>
+                <spark.version>3.0.0</spark.version>
+                <scala.version>2.12.12</scala.version>
+                <scala.compat.version>2.12</scala.compat.version>
+            </properties>
+        </profile>
+        <profile>
+            <id>scala-2.12_spark-3.0.1</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
+            </activation>
+            <properties>
+                <jetty.version>9.4.31.v20200723</jetty.version>
+                <slf4j.version>1.7.16</slf4j.version>
+                <spark.version>3.0.1</spark.version>
+                <scala.version>2.12.12</scala.version>
+                <scala.compat.version>2.12</scala.compat.version>
+            </properties>
+        </profile>
     </profiles>

     <properties>
```

**Review by hallihan — CHANGES_REQUESTED** [review]
A few small suggestions to align profiles between the main project and the sample job pom.xml.  I already ran the updated code against Azure Databricks 5.5, 6.6, 7.2, and 7.4 without any issues.  Thanks a ton for this contribution!

File: src/pom.xml (hunk 3 above)
Comment (dfanesiDB) [reply src/pom.xml:37]: @hallihan thank you for reviewing my changes. I applied all your suggestions.

**Review by hallihan — APPROVED** [review]
